file access error handling

from the discord https://discord.com/channels/860500297297821756/860500297297821759/971911693443686483 as reported by SleepyNetworker

> I'm running into an error when running containerlab. I think it's happening because Clab does not have access to my home folder and it's failing because it cannot access my ~.ssh/ folder.
> Is there a way to configure Clab to not check for keys?
> Here's the error output:
```
➜  ansible-arista-deploy git:(clab) ✗ sudo clab deploy -t clab_topologies/clab-lan-topology.yaml
INFO[0000] Containerlab v0.26.0 started
INFO[0000] Parsing & checking topology file: clab-lan-topology.yaml
INFO[0000] Creating lab directory: /local/mnt/workspace/ansible-arista-deploy/clab-aristalan
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x83151c]

goroutine 1 [running]:
github.com/srl-labs/containerlab/utils.FileExists({0xc0008837a0, 0x10})
    github.com/srl-labs/containerlab/utils/file.go:30 +0x5c
github.com/srl-labs/containerlab/clab.(*CLab).CreateAuthzKeysFile(0xc000351ef0)
    github.com/srl-labs/containerlab/clab/authz_keys.go:38 +0xd7
github.com/srl-labs/containerlab/cmd.deployFn(0x3df4420, {0x24ef0ea, 0x2, 0x2})
    github.com/srl-labs/containerlab/cmd/deploy.go:152 +0x5b6
github.com/spf13/cobra.(*Command).execute(0x3df4420, {0xc000061b00, 0x2, 0x2})
    github.com/spf13/cobra@v1.4.0/command.go:856 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0x3df2b20)
    github.com/spf13/cobra@v1.4.0/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
    github.com/spf13/cobra@v1.4.0/command.go:902
github.com/srl-labs/containerlab/cmd.Execute()
    github.com/srl-labs/containerlab/cmd/root.go:44 +0x25
main.main()
    github.com/srl-labs/containerlab/main.go:10 +0x17
```

--- a/types/types.go
+++ b/types/types.go
@@ -144,7 +144,7 @@ func (node *NodeConfig) GenerateConfig(dst, templ string) error {
 		return err
 	}
 	log.Debugf("node '%s' generated config: %s", node.ShortName, dstBytes.String())
-	f, err := os.Create(dst)
+	f, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0666)
 	if err != nil {
 		return err
 	}

--- a/types/types.go
+++ b/types/types.go
@@ -144,7 +144,7 @@ func (node *NodeConfig) GenerateConfig(dst, templ string) error {
 		return err
 	}
 	log.Debugf("node '%s' generated config: %s", node.ShortName, dstBytes.String())
-	f, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0666)
+	f, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0666) // skipcq: GSC-G302
 	if err != nil {
 		return err
 	}

--- a/utils/file.go
+++ b/utils/file.go
@@ -5,7 +5,6 @@
 package utils
 
 import (
-	"bufio"
 	"errors"
 	"fmt"
 	log "github.com/sirupsen/logrus"
@@ -149,26 +148,6 @@ func ReadFileContent(file string) ([]byte, error) {
 	b, err := ioutil.ReadFile(file)
 
 	return b, err
-}
-
-func ReadFileLines(file string) ([]string, error) {
-	// check file exists
-	if !FileExists(file) {
-		return nil, fmt.Errorf("%w: %s", errFileNotExist, file)
-	}
-	content, err := os.Open(file)
-	if err != nil {
-		return nil, err
-	}
-
-	scanner := bufio.NewScanner(content)
-	scanner.Split(bufio.ScanLines)
-	var result []string
-
-	for scanner.Scan() {
-		result = append(result, scanner.Text())
-	}
-	return result, nil
 }
 
 // ExpandHome expands `~` char in the path to home path of a current user in provided path p.

--- a/utils/file.go
+++ b/utils/file.go
@@ -9,7 +9,6 @@ import (
 	"fmt"
 	log "github.com/sirupsen/logrus"
 	"io"
-	"io/ioutil"
 	"net/http"
 	"os"
 	"path/filepath"
@@ -139,14 +138,8 @@ func CreateDirectory(path string, perm os.FileMode) {
 }
 
 func ReadFileContent(file string) ([]byte, error) {
-	// check file exists
-	if !FileExists(file) {
-		return nil, fmt.Errorf("%w: %s", errFileNotExist, file)
-	}
-
-	// read and return file content
-	b, err := ioutil.ReadFile(file)
-
+	// try to read and return file content, or return an error
+	b, err := os.ReadFile(file)
 	return b, err
 }
 

--- a/utils/file.go
+++ b/utils/file.go
@@ -8,6 +8,7 @@ import (
 	"bufio"
 	"errors"
 	"fmt"
+	log "github.com/sirupsen/logrus"
 	"io"
 	"io/ioutil"
 	"net/http"
@@ -23,7 +24,8 @@ var errHTTPFetch = errors.New("failed to fetch http(s) resource")
 // FileExists returns true if a file referenced by filename exists
 func FileExists(filename string) bool {
 	f, err := os.Stat(filename)
-	if os.IsNotExist(err) {
+	if err != nil {
+		log.Debugf("error while trying to access file %v: %v", filename, err)
 		return false
 	}
 

--- a/utils/file.go
+++ b/utils/file.go
@@ -132,8 +132,9 @@ func CreateFile(file, content string) (err error) {
 // CreateDirectory creates a directory by a path with a mode/permission specified by perm.
 // If directory exists, the function does not do anything.
 func CreateDirectory(path string, perm os.FileMode) {
-	if _, err := os.Stat(path); os.IsNotExist(err) {
-		_ = os.MkdirAll(path, perm)
+	err := os.MkdirAll(path, perm)
+	if err != nil {
+		log.Debugf("error while creating a directory path %v: %v", path, err)
 	}
 }
 

--- a/utils/file.go
+++ b/utils/file.go
@@ -21,7 +21,8 @@ var errNonRegularFile = errors.New("non-regular file")
 var errFileNotExist = errors.New("file does not exist")
 var errHTTPFetch = errors.New("failed to fetch http(s) resource")
 
-// FileExists returns true if a file referenced by filename exists
+// FileExists returns true if a file referenced by filename exists & accessible
+//
 func FileExists(filename string) bool {
 	f, err := os.Stat(filename)
 	if err != nil {

--- a/utils/file.go
+++ b/utils/file.go
@@ -16,7 +16,6 @@ import (
 )
 
 var errNonRegularFile = errors.New("non-regular file")
-var errFileNotExist = errors.New("file does not exist")
 var errHTTPFetch = errors.New("failed to fetch http(s) resource")
 
 // FileExists returns true if a file referenced by filename exists & accessible

